src/goto-programs/remove_function_pointers.cpp

```diff
@@ -338,41 +356,74 @@ void remove_function_pointerst::remove_function_pointer(
   assert(function.id()==ID_dereference);
   assert(function.operands().size()==1);
 
-  const exprt &pointer=function.op0();
+  bool found_functions;
 
-  // Is this simple?
-  if(pointer.id()==ID_address_of &&
-     to_address_of_expr(pointer).object().id()==ID_symbol)
+  const exprt &pointer=function.op0();
+  remove_const_function_pointerst::functionst functions;
+  does_remove_constt const_removal_check(goto_program, ns);
+  if(const_removal_check())
   {
-    to_code_function_call(target->code).function()=
-      to_address_of_expr(pointer).object();
-    return;
+    warning() << "Cast from const to non-const pointer found, only worst case"
+              << " function pointer removal will be done." << eom;
+    found_functions=false;
   }
+  else
+  {
+    remove_const_function_pointerst fpr(
+    get_message_handler(), pointer, ns, symbol_table);
+
+    found_functions=fpr(functions);
 
-  typedef std::list<exprt> functionst;
-  functionst functions;
+    // Consistency checks
+    // Reported optimized function pointer call, but didn't find any functions
+    assert(!found_functions || !functions.empty());
 
-  bool return_value_used=code.lhs().is_not_nil();
+    // Reported didn't optimize function pointer call, but did find some
+    // functions to replace with
+    assert(found_functions || functions.empty());
```

Comment:
I knew there was a way to simplify this, I wanted to use a (non-existent) implication logical operator but you're right it is an if and only if relation so this works.